accepted/2021/experimental-features/experimental-features.md

```diff
@@ -0,0 +1,380 @@
+also declared. For example, if a project has this configuration:
+
+```xml
+<Project>
+
+  <PropertyGroup>
+    <ExperimentalFeatures>StaticInterfaceMembers</ExperimentalFeatures>
+  </PropertyGroup>
+
+</Project>
+```
+
+consuming an assembly with these attributes
+
+```C#
+[assembly: ExperimentalFeature("StaticInterfaceMembers")]
+[assembly: ExperimentalFeature("FancyGC")]
+```
+
+would result in a diagnostic like this:
+
+> error: Assembly 'Foo.Bar' depends on the experimental feature 'FancyGC' which
+> the consuming project has not enabled.
+
+***Note:*** Since F# doesn't support analyzers, we would have to add this
+capability to the compiler itself.
+
+## Q & A
+
+### Why not just have a single on/off switch for experimental features?
+
+We could just have a setting in the project file like this:
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <CanUseExperimentalFeatures>True</CanUseExperimentalFeatures>
+  </PropertyGroup>
+
+</Project>
+```
+
+The problem with that is that once any experimental feature is used, it becomes
+a free-for-all for all experimental features. For example, imagine we ship two
+features:
+
+* Static interface members
+* A new GC mode
+
+Someone who wants to use the new GC mode turns this setting on. Now someone else
+who also works on the same project could now accidentally use static interface
+members, without realizing that that feature is experimental.
+
+This would violate our guiding principle that customers are aware when they
+depend on experimental features.
+
+### Why are we recording experimental features in the assembly?
+
+For two reasons:
+
+1. It can be used for auditing
+2. It is used to enforce transitivity. A project can only depend on binaries
+   that have a subset of experimental features enabled. It doesn't matter
+   whether this binary was a project-to-project reference, a NuGet package, or a
+   raw reference to some checked in file -- we only have to look at the assembly
+   level attributes and we can check which features were enabled when it was
+   compiled.
+
+### What happens when an experimental feature is released?
+
+Several things will happen:
+
+1. We remove the `ExperimentalFeature` attribution from any APIs
+2. We remove any guards from the runtime/compilers that check for the feature
+   flag (of course, compilers are likely still checking language versions, but
+   that's separate)
+
+If someone shipped a binary to nuget.org that depends on an experimental
+feature, consuming it from a .NET version where the feature was release will
+still generate an error unless the consuming project has the feature turned on.
+If we believe this penalizes early adopters too much, we could make sure that
+the .NET SDK knows about which experimental features such that diagnostics for
+now released features can be suppressed.
+
+### Can we make breaking changes in experimental features?
```

Comment:
It's one thing if experiments come and go through the year. But as a platform, if the experiment goes out in an LTS release, a feature might be "experimental" for several years. Some folks may widely deploy apps that use it. Years later, it's time to upgrade and they are blocked because the experimental feature is gone. Yes, I explicitly set the experimental flag, but if the impact is large enough, it can still be hard to pull the API - just ask Windows how often they've been forced to support undocumented API or behaviors.

How can we avoid this "playing chicken" situation? The fact that once you have it in LTS, its available for several years seems to exacerbate it.

Reply:
>  just ask Windows how often they've been forced to support undocumented API or behaviors.

But Windows doesn't have side-by-side and app-local deployment models 😛

Reply:
> Some folks may widely deploy apps that use it

I do not think we can sign up for supporting using experimental features in production. We need to be very clear about that.

If somebody wants to use the experimental features in production, they can only do it on their own risk.

Reply:
I agree - we should be clear that not only does experimental mean “may break or go away” but also “unsupported in production”.

I just wonder whether there is some way we can make it technically hard to use in production, to avoid the hazard on both sides.

Reply:
In our experiences on the compiler we've been able to successfully avoid people falling into this trap by 

1. Being very up front about expectations with the customers. The fact that we call it `<LangVersion>preview</LangVersion>` helps set expectations right from the start. The designs we're discussing here, using `Experimental`, is an even stronger message. 
2. Be aggressive about putting a feature per release into our preview train that isn't fully fleshed out. In part so we can get customer feedback on the feature to help shape it. In other part though to somewhat guarantee that we're going to have "break it" a couple of times during the preview. That regular break pattern helps remind customers that breaks **will** happen in preview mode. If you're a customer who doesn't that then preview is not the experience you're looking for. 
3. Stick to our principals here. We've made a few breaks that didn't go over well with customers but we pointed to the preview documentation and stuck to our design decisions.  

One advantage the compiler team has though is we ship roughly every two months in RTM releases. So we can iterate decently quickly, breaks happen in a relatively short time frame and it keeps customers from getting lulled into a false sense of security. The lead time for LTS won't have this and it may be worth thinking other ways we can help periodically remind customers about the bet they've made.

Reply:
Thought experiment. Would we have any qualms about putting out a servicing release that broke an experimental feature (intentionally or unintentionally)?

Reply:
> I just wonder whether there is some way we can make it technically hard to use in production, to avoid the hazard on both sides.

The plan we have been on in the past is to have the experimental features enabled for preview builds only, and disable the experimental features for the final release. It made it impossible to use the experimental features by accident in production. It would require running preview build in production by accident - if somebody does that, they have bigger problems.

This plan does not allow playing with experimental features unless you have a preview. If we go with the premise that the feature iterations really happens in previews, it does not sound too bad.

Reply:
> This plan does not allow playing with experimental features unless you have a preview. 

I cannot see that in the document -- @terrajobst if that matches your intent, could you clarify? If I understand correctly, just before GA we would push a final change that would disable the opt-in flag for experiments. That would address my concerns.

Reply:
> features enabled for preview builds only, and disable the experimental features for the final release.

To be honest, this plan was not without problems. The boundary between experimental and non-experimental was fuzzy. It is what lead us to creating dotnet/runtimelab repo: https://github.com/dotnet/runtime/issues/35609 . 

I assume that we will still want to keep the more speculative experiments in dotnet/runtimelab even with this proposal. This proposal is for features that we are almost certain to ship, but we are just not able to bake them in one release cycle.

Reply:
I expect runtimelab will also remain the preferred place for experiments that require invasive changes in the shipping code paths. A good fit for this proposal would be some new API that leaves shipping code paths essentially unaffected, or a flag that forces a condition or two.

Reply:
> Thought experiment. Would we have any qualms about putting out a servicing release that broke an experimental feature (intentionally or unintentionally)?

_I think_ it should be fair game to break experimental features (intentionally or unintentionally). I can imagine cases where we'd need to.

Reply:
I have added section about quality bar that should address this point.

Reply:
OK @jkotas note that this is not the plan:

> This plan does not allow playing with experimental features unless you have a preview.

Reply:
In practice, I expect you will want to have a preview installed to play with experimental features. Consider this:

1. We ship a .NET release, the VS/compiler sync-shipping with  the .NET release will work well. 
2. One or two months later, we ship first vNext preview and VS will pick up matching changes. It is safe to assume that the vNext will include some breaking changes in the preview features. It means that the current VS is not going to work great with current stable release. For example, you may get a ton of red squiggles everywhere.
3. It is between hard and impossible to downgrade VS. Your most viable options is going to be to target vNext preview runtimes; or lose the VS bells and whistles.

We need to optimize this on getting feedback and iterating in preview .NET release. @jaredpar made the same point in https://github.com/dotnet/designs/pull/185/#discussion_r590483029 . Latest stable runtime + current VS is likely going to be poor experience most of the time.

With this point of view, I think it would be ok to drop the preview features from the official runtime release. It would be good hygiene and address all issue with accidental dependencies.

Reply:
> We need to optimize this on getting feedback and iterating in preview .NET releases

This also helps with focus and marketing. When we ship a new release, the message coming with it is "go upgrade your apps now". We may not want to dilute it with asking for feedback about preview features at the same time.